src/ansys/sound/core/sound_composer/source_harmonics.py

```diff
@@ -0,0 +1,327 @@
+                str_order_values = str(orders)
+
+            # Order control values.
+            control_values = np.round(control_values, 1)
+            if len(control_values) > 10:
+                str_control_values = (
+                    f"{str(control_values[:5])[:-1]} ... {str(control_values[-5:])[1:]}"
+                )
+            else:
+                str_control_values = str(control_values)
+
+            str_source = (
+                f"'{str_name}'\n"
+                f"\tNumber of orders: {len(orders)}\n"
+                f"\t\t{str_order_values}\n"
+                f"\tControl parameter: {control_name}, "
+                f"{np.round(np.min(control_values), 1)} - "
+                f"{np.round(np.max(control_values), 1)} rpm\n"
+                f"\t\t{str_control_values}"
+            )
+        else:
+            str_source = "Not set"
+
+        # Source control info.
+        if self.is_source_control_valid():
+            str_source_control = (
+                f"{self.source_control.control.name}\n"
+                f"\tMin: {self.source_control.control.data.min()}\n"
+                f"\tMax: {self.source_control.control.data.max()}\n"
+                f"\tDuration: "
+                f"{self.source_control.control.time_freq_support.time_frequencies.data[-1]} s"
+            )
+        else:
+            str_source_control = "Not set/valid"
+
+        return f"Harmonics source: {str_source}\nSource control: {str_source_control}"
+
+    @property
+    def source_control(self) -> SourceControlTime:
+        """Harmonics source control.
+
+        Contains the control parameter values over time.
+        """
+        return self.__source_control
+
+    @source_control.setter
+    def source_control(self, source_control: SourceControlTime):
+        """Set the source control."""
+        if not (isinstance(source_control, SourceControlTime) or source_control is None):
+            raise PyAnsysSoundException(
+                "Specified source control object must be of type ``SourceControlTime``."
+            )
+        self.__source_control = source_control
+
+    @property
+    def source_harmonics(self) -> FieldsContainer:
+        """Harmonics source data, as a DPF fields container.
+
+        The harmonics source data consists of a series of orders whose levels depend on RPM.
+        """
+        return self.__source_harmonics
+
+    @source_harmonics.setter
+    def source_harmonics(self, source: FieldsContainer):
+        """Set the harmonics source data, from a DPF fields container."""
+        if source is not None:
+            if not isinstance(source, FieldsContainer):
+                raise PyAnsysSoundException(
+                    "Specified harmonics source must be provided as a DPF fields container."
+                )
+
+            if len(source) < 1:
+                raise PyAnsysSoundException(
+                    "Specified harmonics source must contain at least one order."
+                )
+
```

Comment:
@ansaminard Replace next for by:

nb_orders = len(self.source_harmonics[0].time_freq_support.time_frequencies.data) # nb orders
if (nb_orders < 1)
  display error

for f_order in source:
    if len(f_order.data) != nb_orders:
        raise PyAnsysSoundException(
                        "Each set of order in the specified harmonics source must contain the same number of orders"
                    )

Reply:
@anshlachamb I've reworked a bit the tests you suggested, such that the error messages make more sense to the user. But now we should cover all possible cases